[nccl] Remove lock for nccl collective launch for 2.0+

xw285cornell opened the pull request:
Summary: It looks nccl 2.0+ no longer needs a lock to avoid being called concurrently with cudaFree.

Test Plan: sandcastle + OSS CI

Differential Revision: D44514446

suo commented:
@kwen2501 mind taking a look at this one :)

malfet commented:
Adding `ciflow/trunk` to test distributed changes. 
Would be good to add a C++ unittest specifically for this types of race conditions

kwen2501 reviewed (approved):
LGTM. Thanks!

nit: I don't think we are still supporting NCCL 1.x. Please feel free to remove all `#if .. #endif` together with that lock :)

xw285cornell committed:
[nccl] Remove lock for nccl collective launch for 2.0+ (#97904)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/97904

It looks nccl 2.0+ no longer needs a lock to avoid being called concurrently with cudaFree.

Test Plan:
sandcastle + OSS CI

added a test with cudaFree + nccl

{F942888881}

Reviewed By: kwen2501

Differential Revision: D44514446

fbshipit-source-id: fb4aebd91b06a90dd2f403ffb00752598b5720a9
